docs/proposals/cross-mesh-ux.md

```diff
@@ -0,0 +1,229 @@
+# Cross-mesh communication UX
+
+_Issue_: https://github.com/kumahq/kuma/issues/3323
+
+This proposal allows the operators of one `Mesh` to expose services to other `Meshes`
+via `MeshGateway`, the same interface for exposing to external traffic.
+
+The proposal addresses UX both for _exposing_ and the _consuming_
+meshes and only HTTP/S services. TCP will be addressed in a later proposal.
+
+## Design
+
+In the following examples, the mesh exposing its services will be `expose` and
+the mesh consuming another mesh's services will be `consume`.
+
+<details>
+<summary>Recap: <code>MeshGateway</code> configuration</summary>
+
+```
+type: MeshGateway
+mesh: expose
+name: edge-gateway
+selectors:
+- match:
+    kuma.io/service: edge-gateway
+conf:
+  listeners:
+  - port: 8080
+    protocol: HTTP
+    hostname: foo.example.com
+    tags:
+      port: "8080"
+---
+type: MeshGatewayRoute
+mesh: expose
+name: edge-gateway-route
+selectors:
+  - match:
+      kuma.io/service: edge-gateway
+conf:
+  http:
+    rules:
+      - matches:
+          - path:
+              match: PREFIX
+              value: /api
+        backends:
+          - destination:
+              kuma.io/service: server
+```
+
+</details>
+
+### `expose` UX
+
+The main UX feature will be a mechanism to configure a
+`MeshGateway` to permit traffic from services in one
+`Mesh` to reach the services behind the `MeshGateway`.
+The configuration will make it possible to expose `MeshGatewayRoute`s depending
+on cross-mesh traffic's source `Mesh` and source _tags_.
+
+First of all, we require users to explicitly enable cross-mesh traffic on
+the `MeshGateway` listener as well as mTLS on any meshes doing cross-mesh
+traffic.
+This option will allow us to configure the corresponding Envoy listener
+with all mesh CAs.
+
+```
+type: MeshGateway
+mesh: expose
+name: edge-gateway
+selectors:
+- match:
+    kuma.io/service: edge-gateway
+conf:
+  listeners:
+  - port: 8090
+    protocol: HTTPS
+    hostname: expose.mesh # optional
+    tags:
+      name: mesh-listener
+    tls:
+      mode: TERMINATE
+      options:
+        crossMesh: true # new option
+```
+
+There are two ways we can enable configuring more specific access rules.
+
+#### `MeshGatewayTrafficPermission`
```

Comment:
Do you think it will be possible to merge it with [MeshTrafficPermission](https://github.com/kumahq/kuma/issues/4222) in the future ?

Reply:
Hmm. One restricts traffic to services and one traffic to a `MeshGateway`. But it looks like there may be some definite overlap with the `http` options in `MeshTrafficPermission`, it's definitely something we need to figure out. There's already some overlap between `MeshGatewayRoute` and `TrafficRoute`.

Reply:
I'm conflicted.

On the one hand, we have a proliferation of CRs and that can in general lead to confusion about when to use which. If there _is_ a common abstraction we should find it and make it explicit.

On the other hand, in this case it feels OK to separate permissions for _entering_ the mesh (via a `MeshGateway`) from permissions for traffic between services (between `Dataplane`s) inside a mesh? A `MeshGateway` `Dataplane` is genuinely _different_ than a normal `Dataplane`, at least it is right now in that it's configured by the user differently and we configure Envoy differently in the implementation. But maybe I would be more comfortable with it if `TrafficRoute` and `MeshGatewayRoute` weren't different resources already and/or we had a common abstraction between `MeshGateway` and `Dataplane`s?

Reply:
After the convo with @lobkovilya I agree I think it should be possible to unify them.

> For MeshGatewayRoute it was a bit different because we had to adjust a lot for Gateway API, but in the future this might also make sense to merge them maybe.

Yes, I think it should be possible to make our underlying resource strictly more powerful than Gateway API so HTTPRoute can be transformed into it.

Reply:
OK, the proposal now references the `MeshTrafficPermission` tracking issue.